zerver/actions/users.py

```diff
@@ -180,89 +110,114 @@ def do_delete_user_preserving_messages(user_profile: UserProfile) -> None:
-        replacement_user.recipient = personal_recipient
-        replacement_user.save(update_fields=["recipient"])
-
-        # Uses index: zerver_message_realm_sender_recipient (prefix)
-        Message.objects.filter(realm_id=realm.id, sender=temp_replacement_user).update(
-            sender=replacement_user
+
+        overwrite_data = model_to_dict(
+            temp_replacement_user,
+            exclude=[
+                "id",
+                "date_joined",
+                "realm",
+                "recipient",
+                "uuid",
+                "last_login",
+                # These come from Django and aren't used by Zulip. We need to exclude them as they are m2m fields
+                # and as such, would break the code below.
+                "groups",
+                "user_permissions",
+            ],
         )
+
+        UserProfile.objects.filter(id=user_profile.id).update(**overwrite_data)
+        user_profile.refresh_from_db()
+
+        update_all_subscriber_counts_for_user(user_profile=user_profile, direction=-1)
+        Subscription.objects.filter(
+            user_profile=user_profile, recipient__type=Recipient.STREAM
+        ).delete()
+
         Subscription.objects.filter(
-            user_profile=temp_replacement_user, recipient__type=Recipient.DIRECT_MESSAGE_GROUP
-        ).update(user_profile=replacement_user, is_user_active=replacement_user.is_active)
-        temp_replacement_user.delete()
+            user_profile=user_profile, recipient__type=Recipient.DIRECT_MESSAGE_GROUP
+        ).update(is_user_active=False)
+
+        fks_to_delete: list[tuple[Any, str]] = [
+            (UserMessage, "user_profile"),
+            (ExternalAuthID, "user"),
+            (SavedSnippet, "user_profile"),
+        ]
+        for table, field_name in fks_to_delete:
+            table.objects.filter(**{field_name: user_profile}).delete()
+
+        # These audit logs can carry personal information that we want to purge
+        # in the process of user data deletion - and are not that important
+        # to preserve.
+        audit_log_event_types_for_deletion = [
+            AuditLogEventType.USER_EMAIL_CHANGED,
+            AuditLogEventType.USER_FULL_NAME_CHANGED,
+            AuditLogEventType.USER_SETTING_CHANGED,
+        ]
+        # TODO: We could be a bit more conservative here, and just purge .extra_data instead of .delete() the objects?
+        # That's what contains personal data. Could add a scrubbed: bool field to RealmAuditLog and mark it True
+        # for these.
+        RealmAuditLog.objects.filter(
+            modified_user=user_profile, event_type__in=audit_log_event_types_for_deletion
+        ).delete()
+
+        if delete_messages:
+            # TODO: should this be moved to deferred_work queue?
+            # TODO: do we want to delete personal messages sent *TO* this user?
+            do_delete_messages_by_sender(user_profile)
```

Comment:
We should probably discuss in chat.zulip.org for visibility, My initial thought would be that the specification of "deleting the messages you've sent" does not include deleting messages others' sent you; those aren't yours to delete.